contributors/design-proposals/runas-groupid.md

```diff
@@ -0,0 +1,191 @@
+     // PodSecurityContext, the value specified in SecurityContext takes precedence
+     // for that container.
+     // +optional
+     RunAsUser *int64
+     // The GID to run the entrypoint of the container process.
+     // Defaults to group specified in image metadata if unspecified.
+     // May also be set in PodSecurityContext.  If set in both SecurityContext and
+     // PodSecurityContext, the value specified in SecurityContext takes precedence.
+     // +optional
+     RunAsGroup *int64
+
+    .....
+ }
+```
+
+#### PodSecurityPolicy
+
+PodSecurityPolicy defines strategies or conditions that a pod must run with in order to be accepted
+into the system. Two of the relevant strategies are RunAsUser and SupplementalGroups. We introduce 
+a new strategy called RunAsGroup which will support the following options:
+- MustRunAs
+- MustRunAsNonRoot
+- RunAsAny
+
+```
+// PodSecurityPolicySpec defines the policy enforced.
+ type PodSecurityPolicySpec struct {
+     //Other fields not shown for brevity
+    ..... 
+  // RunAsUser is the strategy that will dictate the allowable RunAsUser values that may be set.
+  RunAsUser RunAsUserStrategyOptions
+  // SupplementalGroups is the strategy that will dictate what supplemental groups are used by the SecurityContext.
+  SupplementalGroups SupplementalGroupsStrategyOptions
+
+
+  // RunAsGroup is the strategy that will dictate the allowable RunAsGroup values that may be set.
+  RunAsGroup RunAsGroupStrategyOptions
+   .....
+}
+
+// RunAsGroupStrategyOptions defines the strategy type and any options used to create the strategy.
+ type RunAsUserStrategyOptions struct {
+     // Rule is the strategy that will dictate the allowable RunAsGroup values that may be set.
+     Rule RunAsGroupStrategy
+     // Ranges are the allowed ranges of gids that may be used.
+     // +optional
+     Ranges []GroupIDRange
+ }
+
+// RunAsGroupStrategy denotes strategy types for generating RunAsGroup values for a
+ // SecurityContext.
+ type RunAsGroupStrategy string
+ 
+ const (
+     // container must run as a particular gid.
+     RunAsGroupStrategyMustRunAs RunAsGroupStrategy = "MustRunAs"
+     // container must run as a non-root gid
+     RunAsGroupStrategyMustRunAsNonRoot RunAsGroupStrategy = "MustRunAsNonRoot"
+     // container may make requests for any gid.
+     RunAsGroupStrategyRunAsAny RunAsGroupStrategy = "RunAsAny"
+ )
+```
+
+## Behavior
+
+Following points should be noted:
+
+- `FSGroup` and `SupplementalGroups` will continue to have their old meanings and would be untouched.  
+- The `RunAsGroup` In the SecurityContext will override the `RunAsGroup` in the PodSecurityContext.
+- If no `RunAsGroup` is provided in the PodSecurityContext and SecurityContext, the Group provided 
+  in the Docker image will be used.
+- If no `RunAsGroup` is provided in the PodSecurityContext and SecurityContext, and none in the image,
+  the container will run with primary Group as root(0).
+
```

Comment:
might note here that non-numeric groups will cause an error like the behavior for `runAsUser` in the `MustRunAsNonRoot` case.

Reply:
@pweil- the api fields only take numeric values here.

Reply:
@krmayankk `MustRunAsNonRoot` (for `runAsUser`) includes a kubelet check that parses the value from the dockerfile which main contain non-numeric values.  If it is a non-numeric value we error.  I think you want the same thing here.  Please correct me if I'm wrong.

Reply:
@pweil-  You are right about the check. Some additional context.  In SecurityContext RunAsNonRoot is a bool which if set to true, the docker image user is verified here https://github.com/kubernetes/kubernetes/blob/b32639f9e7d47d0f69bb41b5535803ae4c9d835a/pkg/kubelet/kuberuntime/security_context.go#L81
 It seems we will have to add an additional check to verify both user and group in that file. Also the non numeric verification you mention is here 
https://github.com/kubernetes/kubernetes/blob/9737efccf0e026e2e2437aa31c4365577d628bd5/pkg/kubelet/kuberuntime/kuberuntime_container.go#L186
 The check you mention is against SecurityContext.RunAsNonRoot and not against PSP.MustRunAsNonRoot policy but i am sure it somehow plumbs through. I will add an additional bullet to what you mention.
@pmorie @pweil-  do you both agree that RunAsNonRoot boolean in SecurityContext should apply to both uid and gid and not just user ?

Reply:
> The check you mention is against SecurityContext.RunAsNonRoot and not against PSP.MustRunAsNonRoot policy but i am sure it somehow plumbs through

the strategy sets the boolean to true during admission

> @pmorie @pweil- do you both agree that RunAsNonRoot boolean in SecurityContext should apply to both uid and gid and not just user ?

hrm, that's sticky.  They are two different strategies so they could theoretically be set to different values, a single boolean won't be able to express that condition to allow the kubelet to do the check correctly

Reply:
In that case we would need another boolean `RunAsNonRootGroup` in addition to `RunAsNonRoot` which will continue to apply to uid. I will add that to the proposal

Reply:
@pmorie can you raise your concerns about not adding RunAsNonRootGroup here. @pweil- @smarterclayton i have updated the doc to also say we need `RunAsNonRootGroup` in addition to `RunAsNonRoot` similar to PSP supporting two different strategies  for Group and User. RunAsNonRoot in SecurityContext and PodSecurityContext is specific to uid and hence we need to add one one more bool RunAsNonRootGroup for gid. PTAL and comment here

Reply:
> @krmayankk MustRunAsNonRoot (for runAsUser) includes a kubelet check that parses the value from the dockerfile which main contain non-numeric values. If it is a non-numeric value we error.

@krmayankk @pweil- I don't see where this lead to an error. Could you look at the code and prove that I'm wrong?

I see that:
1) in the `generateContainerConfig` we're [getting uid or username from a Dockerfile](https://github.com/kubernetes/kubernetes/blob/9737efccf0e026e2e2437aa31c4365577d628bd5/pkg/kubelet/kuberuntime/kuberuntime_container.go#L181)
2) we're [invoking `verifyRunAsNonRoot` only when uid is not nil](https://github.com/kubernetes/kubernetes/blob/9737efccf0e026e2e2437aa31c4365577d628bd5/pkg/kubelet/kuberuntime/kuberuntime_container.go#L185-L192). Otherwise, `username` is set to a non-numeric user name and we only print a warning
3) later, we're executing `generateLinuxContainerConfig` that is invoking `determineEffectiveSecurityContext`. [The latter is only using `uid` or `username`](https://github.com/kubernetes/kubernetes/blob/9737efccf0e026e2e2437aa31c4365577d628bd5/pkg/kubelet/kuberuntime/security_context.go#L42-L48) and doesn't do any check whether username is allowed or it's not.

Reply:
Yes, it looks like when non-numeric support was added [here, I think](https://github.com/kubernetes/kubernetes/commit/99ee3f4b762d5a9f5d6736b62ffeba383b3d7beb#diff-08c8f9ff3cc05ba7009dc6a00cba8fa7R81) it was added as a warning.  I feel like this may be a bug though.

Reply:
@Random-Liu could you comment on the aforementioned commit? Was it made by mistake or by purpose?

Reply:
@Random-Liu ping ^^^^